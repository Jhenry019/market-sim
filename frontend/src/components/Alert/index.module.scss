@import '../../scss/root';
@import '../../scss/mixins/alert.scss';


.alert {
  transform: translate(50%, -100%) !important;
  transition: 0.2s transform;
}

.error {
  @include alert;

  color: var(--red-4);
  background-color: var(--red-1);
}

.success {
  @include alert;

  color: var(--green-3);
  background-color: var(--green-1);
}

.visible {
  transform: translate(50%, 30%) !important;
}