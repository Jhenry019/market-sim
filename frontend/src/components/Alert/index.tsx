import React, { useState, useEffect, useContext } from 'react';
import classNames from '../../utils/classNames';
import { AlertContext } from '../../context/alert';
import styles from './index.module.scss';


export interface AlertProps {
  /** The message to be rendered inside the alert */
  children?: string;
  /** The type of alert being delivered */
  Type?: "success" | "error";
  /** A selector to be used in testing environments */
  dataTestId?: string;
};

export default function Alert({
  Type = "success", 
  children = "",
  dataTestId
}: AlertProps): JSX.Element 
{
  const [visible, setVisible] = useState<boolean>(children !== '');
  const context = useContext(AlertContext);

  useEffect(() => {
    /**
     * The Alert component should hide itself
     * after 3 seconds. The component will still
     * be mounted, but it will be given a className
     * that visually hides it from the user's view.
     */
    if (children) {
      setVisible(true);

      const hide = setTimeout(() => {
        setVisible(false);
      }, 3000);

      return function () {
        clearTimeout(hide);
      }
    }
  }, [context]);

  return (
    <div 
      className={classNames(
        styles.alert,
        styles[`${Type}`],
        visible && styles.visible
      )}
      data-testid={dataTestId}
    >
      <div className={styles.alertMssg}>
        {children}
      </div>
    </div>
  );
}