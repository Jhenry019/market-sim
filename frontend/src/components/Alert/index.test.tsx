import React from 'react';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Alert from '.';


beforeEach(() => {
  /** The Alert component uses 'setTimeout'
   *  in its useEffect hook. This ensures
   *  that 'setTimeout' gets mocked by
   *  jest.
   */
  jest.useFakeTimers();
});

afterEach(() => {
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
});

it('renders an Alert with a message', () => {
  render(
    <Alert Type="success" dataTestId="alert">
      Successfully purchased 2 shares of ABC
    </Alert>
  );

  const alert: HTMLElement = screen.getByTestId("alert")
  expect(alert.textContent)
    .toBe("Successfully purchased 2 shares of ABC");

  expect(alert.classList).toContain('success');
  expect(alert.classList).toContain('visible');
  expect(alert).toMatchSnapshot();
});

it('renders an Alert which hides itself after 3 seconds', () => {
  act(() => {
    render(
      <Alert Type="success" dataTestId="alert">
        Successfully sold 1 share of ABC
      </Alert>
    );
    // Allow the timer inside the Alert component to run
    jest.runAllTimers();
  });

  expect(screen.getByTestId("alert").classList)
    .not.toContain('visible')
});

it('renders an Alert with an error className', () => {
  render(
    <Alert Type="error" dataTestId="alert">
      Not enough shares of ABC
    </Alert>
  );
  
  expect(screen.getByTestId("alert").classList)
    .toContain('error');
});