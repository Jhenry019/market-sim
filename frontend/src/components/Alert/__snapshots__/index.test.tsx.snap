// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an Alert with a message 1`] = `
<div
  class="alert success visible"
  data-testid="alert"
>
  <div
    class="alertMssg"
  >
    Successfully purchased 2 shares of ABC
  </div>
</div>
`;
